Sources/Commands/SwiftPackageTool.swift

```diff
@@ -966,3 +1051,207 @@ fileprivate func logPackageChanges(changes: [(PackageReference, Workspace.Packag
+    }
+}
+
+fileprivate func makePackage(fileSystem: FileSystem,
+                             configPath: AbsolutePath,
+                             packageName: String?,
+                             mode: MakePackageMode,
+                             packageType: InitPackage.PackageType?,
+                             packageTemplate: String?) throws {
+    
+    guard let cwd = fileSystem.currentWorkingDirectory else {
+        throw InternalError("Could not find the current working directroy.")
+    }
+    
+    guard !(packageType != nil && packageTemplate != nil) else {
+        throw InternalError("Can't use --type in conjunction with --template")
+    }
+    
+    let name: String
+    let destinationPath: AbsolutePath
+    let templateHomeDirectory = configPath.appending(components: "templates", "new-package")
+    
+    var foundTemplate = false
+    var templateToUse = ""
+    
+    if let templateName = packageTemplate {
+        guard fileSystem.exists(templateHomeDirectory.appending(component: templateName)) else {
+            throw InternalError("Could not find template folder: \(templateHomeDirectory.appending(component: templateName))")
+        }
+        
+        templateToUse = templateName
+        foundTemplate = true
+    } else {
+        // Checking if a default template is present
+        if fileSystem.exists(templateHomeDirectory.appending(components: "default")) {
+            templateToUse = "default"
+            foundTemplate = true
+        }
+    }
+    
+    switch mode {
+    case .initialize:
+        name = packageName ?? cwd.basename
+        destinationPath = cwd
+    case .create:
+        name = packageName!
+        try localFileSystem.createDirectory(cwd.appending(component: name))
+        destinationPath = cwd.appending(component: name)
+    }
+    
+    let packageTemplate: InitPackage.PackageTemplate
+    if foundTemplate && useNewBehaviour == .new {
+        try fileSystem.getDirectoryContents(templateHomeDirectory.appending(component: templateToUse)).forEach {
+            print("Copying \($0)")
+            try copyTemplate(fileSystem: fileSystem, sourcePath: templateHomeDirectory.appending(components: templateToUse, $0), destinationPath: destinationPath, name: name)
+        }
+        return
+    } else {
+        // These are only needed in the event that --type was not used when creating a package
+        // otherwise if --type is used packageType will not be nill
+        let defualtType = mode == .initialize ? InitPackage.PackageType.library : InitPackage.PackageType.executable
+        packageTemplate = try InitPackage.PackageTemplate(template: getSwiftPMDefaultTemplate(type: packageType ?? defualtType))
+    }
+
+    let initPackage = try InitPackage(name: name, destinationPath: destinationPath, packageTemplate: packageTemplate)
+    initPackage.progressReporter = { message in
+        print(message)
+    }
+    try initPackage.writePackageStructure()
+}
```

Comment:
if I read this correctly, I dont think this follow the logic defined in the evolution proposal.

I suggest restructuring the code in the following way:

1. separate `makePackage` into 2 functions, one for the "legacy behavior" and one for the "new" one.
2. from the Init command, call the legacy one when mode == .legacy and the new one when mode == .new
3. initPackage.writePackageStructure should wither take a mode argument, or also be split into legacy/new since per the design its supposed to work a bit differently in legacy vs.  new mode (e.g. not create certain files)